### ISSUE TITLE
[closeness.py] Should the Graph G be reversed? [edited]

### ISSUE THREAD

**Issue by luanamarinho** [networkx/networkx#3463]
Hi all,

I have a few comments on the closeness_centrality function:

1. In the block comment, " **average** shortest path distance **to** `u` over all `n-1` reachable nodes" should be corrected. It's the reciprocal of the **sum** of the shortest path distances **from** u to all n-1 reachable nodes (given that u can reach itself in the distance methods)

> Closeness centrality [1]_ of a node `u` is the reciprocal of the
    **average shortest path distance to `u` over all `n-1**` reachable nodes.

2. According to the definition of the variable path_length, which uses either single_source_dijkstra_path_length or single_source_shortest_path_length, the graph input G **should not be reversed**. Hence the command line `G = G.reverse()` should be erased.

3. The notes in https://networkx.github.io/documentation/networkx-1.10/reference/generated/networkx.algorithms.centrality.closeness_centrality.html should be revised.

Regards,
Luana

**Comment by nandahkrishna** [comment]
I believe I can offer an explanation here, hope you find it helpful.
1. The block comment, I think, is right in saying that it is the reciprocal of the average shortest path distance. There are `n-1` reachable nodes from `u` meaning we have `n-1` shortest paths from `u` to these nodes. The average length over all these shortest paths would be the sum of their lengths, divided by `n-1`. Taking this quantity's reciprocal, we get the formula for closeness centrality. So I believe the definition is correct.
2. The block comment also describes that in the case of a directed graph, the inward closeness value is calculated. To use the single source path functions, we reverse the graph so that all inward edges become outward edges. This helps us get the value for inward closeness centrality. For getting the outward closeness, one simply has to pass `G.reverse()` as input to negate the effect of the function's own call of reverse.
3. This seems to be the documentation of NetworkX 1.10, which is not the latest version. The latest stable version is NetworkX 2.3 and 2.4 is in development. I believe their documentation is consistent with the descriptions in the block comments in the function. You could check here (2.3): https://networkx.github.io/documentation/stable/reference/algorithms/generated/networkx.algorithms.centrality.closeness_centrality.html#networkx.algorithms.centrality.closeness_centrality

**Comment by luanamarinho** [comment]
Hi @nandahkrishna,

Thanks for the response, but I disagree in all three points:

1. The sum of the distances from `u` o the `n - 1` reachable nodes divided by `n-1` is a normalized sum. I agree that, mathematically, it equals the average, but that's not quite the purpose. I guess the word choice is not the most appropriate. But, well, "details".
2.  The single source path functions returns the distance **from** node `u` to all `n-1` reachable nodes. The outward distances. Therefore, reversing the the graph gives wrong results. Try to compare the function's results with the values given by SNA softwares such as Cytoscape. I did this comparison and found them to be different. When I decided not to the reverse the graph, the results were the same.

The single_source_shortest_path_length and single_source_dijkstra_path_length methods compute the distances **from** `u` to the reachable nodes (including itself). So the lines below will the the wrong result:
`if G.is_directed():
       G = G.reverse()`
`G.reverse` gives, in fact, the inward, not the desired outward distances.

Please see:
https://github.com/networkx/networkx/blob/master/networkx/algorithms/centrality/closeness.py
https://networkx.github.io/documentation/networkx-1.10/reference/generated/networkx.algorithms.shortest_paths.unweighted.single_source_shortest_path_length.html
https://networkx.github.io/documentation/networkx-1.11/reference/generated/networkx.algorithms.shortest_paths.weighted.single_source_dijkstra_path_length.html

3. Quoting the note on the webpage: 

> The closeness centrality is normalized to (n-1)/(|G|-1) where n is the number of nodes in the connected part of graph containing the node

`n` is not the number of nodes in the connected component in which `u` can be found. Suppose a not strongly connected graph and a node, say, `L`. `L` can be itself a single component but reach node, say, `M`. This definition would give zero, incorrectly. 

Cheers!

**Comment by nandahkrishna** [comment]
It's probably true that the terminology may not be entirely right. I think @dschult may be able to offer a better explanation in these points than I can. 

However, I do not still understand point 2. In NetworkX, we decide that **closeness is calculated based on incoming paths** to the node in directed graphs. This means that we require using every other point as a source and the destination is `u`. But by **reversing**, all **incoming paths become outgoing paths**. This **allows us to use `u` as a source in single source shortest paths, to get all paths to other nodes**. This is **equivalent to getting all paths to `u` from other nodes, in a non-reversed graph**. 

While I have not used the other software, I believe they might use outward closeness as opposed to inward closeness used here. Once again, I may be wrong. If you could offer a slightly more detailed explanation, I'd be grateful. Thanks!

**Comment by dschult** [comment]
It seems that @nandahkrishna has correctly answered your comments 1), 2) and 3). 
Whether you want to believe those answers is another matter. 
Closeness is defined as inward distance. 
It is in fact the reciprocal of the average (though certainly that could be described in other ways).
The definition of "component" takes care of your concerns about ```n```.

The documentation describes accurately what the code does. We didn't just make up the algorithm. Follow the citations in the documentation to see more information about definitions, etc. 
Finally, it would be helpful if you approach the issue without just stating that everything is wrong.
Investigate further to see what we did and why we might have done that. Perhaps your notion of closeness is not the same as ours.

**Comment by luanamarinho** [comment]
Hi all,
Well, it seems this debate is taking the wrong direction. @dschult, thanks for the advice regarding my approach to the matter. I didn't mean to be rude and never assumed you've just made things up. But perhaps my way of questioning the function was too blunt, so I apologize to @nandahkrishna and the rest of the team - that put in a lot of time and effort in keeping this lib up.
I'd love to continue this debate, in more friendly terms, if that's ok.
Cheers!

**Comment by luanamarinho** [comment]
@nandahkrishna @dschult  Again, please don't take my questions as a sign of rudeness. It's not my intention at all to criticise anyone. I really, really just want to understand what the lib does. Now the points:

1. Fair enough. It's a silly debate over terminology, and it is the average of the sum of the distances after all. Agreed.

2. I took the `G.reverse()` to be not quite ok, because I'm following the definition of [outward] closeness centrality given by Freeman [1978], as well as the implementations in Cytospace, Gephi and Neo4j. So if the networkx lib intends to give the inward closeness centrality, that is, considering the distances **from** the `n-1` reachable nodes of `G` **to** a given node `u`, maybe it would be nice to make it crystal clear, so people who does SNA know what they're getting as a result. If you think the documentation is already clear enough on this point, then my bad.

https://www.sciencedirect.com/science/article/pii/0378873378900217
https://github.com/gephi/gephi/wiki/Closeness-Centrality
https://neo4j.com/blog/graph-algorithms-neo4j-closeness-centrality/

3. The code below suggests that the definition of component does not take care of my concern, as I was able to understand it. Forgive me if there's an issue in my interpretation.

`import networkx as nx`
`G = nx.DiGraph()`
`edges = [('D','E'), ('E','D'), ('C','D'), ('E','C'), ('B','C'), ('B','E'), ('A','E'), ('A','B'), ('D','B'), ('C','A'),
         ('G','A'), ('G', 'J'), ('H','G'), ('I','G'), ('F','G'), ('I','H'), ('H','I'), ('I','J'), ('I', 'F'), ('J', 'F'),
        ('A','N'), ('N','O'), ('O','K'), ('K','M'), ('L','M'), ('N','L'), ('K','L'), ('O','L'), ('O','J'), ('J','O')]`
`G.add_edges_from(edges)`
`# %matplotlib notebook`
`# nx.draw_networkx(G)`
`nx.is_strongly_connected(G)`
`# False`
`print(sorted(nx.strongly_connected_components(G)))`
`# [{'M'}, {'L'}, {'K'}, {'O', 'F', 'J', 'D', 'G', 'C', 'B', 'N', 'A', 'E'}, {'H', 'I'}]`
`#closeness centrality`
`print(nx.closeness_centrality(G, 'L'))`
`# 0.32625482625482627 [and not zero]`

Once again, I apologize for my bluntness. I love this lib, honestly!

Cheers!

**Comment by nandahkrishna** [comment]
@luanamarinho no issues at all! Happy to be part of a constructive debate and I understand everyone has different use cases. Glad you find NetworkX to be useful, it's definitely a community effort that enriches packages like these and your doubts are definitely understandable. I will get back to you on this particular point when I've got it cleared up myself. Thanks!

**Comment by ericmjl** [comment]
@luanamarinho FYI I edited your code block jus so that it's formatted correctly. Hope you don't mind :smile:.

**Comment by luanamarinho** [comment]
Not at all, @ericmjl! Thanks for that : )

**Comment by dschult** [comment]
Those are very helpful links and also example code @luanamarinho 
:)

**Comment by dschult** [comment]
I've looked at all three links provided which discuss closeness centrality. This list contains the article by Freeman that we reference in the docs. I am not physically close to my copy of Wasserman and Faust to check that reference, but if I recall correctly, that reference is about how to handle graphs that are not connected; still it is possible they discuss directed graphs.

[Freeman](https://www.sciencedirect.com/science/article/pii/0378873378900217) doesn't consider directed graphs.

[Gephi](https://github.com/gephi/gephi/wiki/Closeness-Centrality) refers to Wikipedia for the definition of closeness centrality and that page clearly defines it as inward distances. I tried to find where in the source code they compute closeness and was not successful.

[neo4j](https://neo4j.com/blog/graph-algorithms-neo4j-closeness-centrality/) also does not discuss directed graphs, though the language clearly uses the words "distance from a node to all other nodes". 

The [Wikipedia definition of closeness centrality](https://en.wikipedia.org/wiki/Closeness_centrality) clearly uses inward closeness as the definition, referring to 2 papers from 1950 and 1966. I have not tracked down those papers.

I have not been able to find a commonly used definition of closeness centrality that uses outward distance from a node. I have found a Wikipedia article that uses inward distance to a node as well as a neoj4 blog that talks about outward distance from a node.  

@luanamarinho, are you able to find more definitive references for which way this is best defined? Otherwise I'm inclined to continue to use the Wikipedia definition that Gephi cites. It would be best to lean on refereed published works, but even code is helpful here... :}
